packages/rpc-provider/src/ws/index.ts

```diff
@@ -334,12 +341,18 @@ export class WsProvider implements ProviderInterface {
   }
 
   #onSocketClose = (event: CloseEvent): void => {
+    const error = new Error(`disconnected from ${this.#endpoints[this.#endpointIndex]}: ${event.code}:: ${event.reason || getWSErrorString(event.code)}`);
+
     if (this.#autoConnectMs > 0) {
-      l.error(`disconnected from ${this.#endpoints[this.#endpointIndex]}: ${event.code}:: ${event.reason || getWSErrorString(event.code)}`);
+      l.error(error.message);
     }
 
     this.#isConnected = false;
     this.#emit('disconnected');
+    // reject all hanging requests
+    eraseRecord(this.#handlers, (handler) => handler.callback(error, undefined));
+    eraseRecord(this.#waitingForId);
+    eraseRecord(this.#subscriptions);
```

Comment:
This all depends. If auto-reconnect, you actually expect it not to fail. Subscriptions especially - in that case it should be re-created when re-connected.

The waiting for Id & single-shot is problematic, since they won't receive an id.

Reply:
even auto reconnected, we would need to re-fire the requests, will it worth all the efforts and ok that we hide these details from user? 
Also, who knows how long it might takes to reconnect, in the end we have to chose between fast fail or add timeout guards

Reply:
This is a major breaking change - the API since pre-1.0 always re-connects subscriptions. So along with it operating differently, it also means that codes needs to be rewritten to cater for the new behavior.

Reply:
Oh, I didn't realize that, will check the code in the morning, this pr is only to prevent hanging promises, don't mean to change the behavior

Reply:
So I believe it is fine for the other 2 cases, i.e. the non-connected subs.